### PR TITLE
Datamining

### LINKED ISSUE #2: Bundesliga Data
Hook into the [Bundesliga API](https://rapidapi.com/heisenbug/api/bundesliga-live-scores) and get some basic datasets that we can play around with.

### PR DESCRIPTION
**Pull request by maxUser**
# Description
* created `get_html.py` which uses `requests` module to get raw html
* created `fussballdatenpunktde_matchday_results.py` to scrape basic matchday results data
* created `run.py` to run both files in succession

Resolves #2 

## Special Notes
* now taking requests for next data to scrape!

## Expected Version Bump
- [ ] Major
- [X] Minor
- [ ] Patch
- [ ] None

# Checklist:
- [X] I have run `black .`
- [X] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: the_mines/data/fussballdaten_2019_2020_1.html [hunk 1]
```diff
@@ -0,0 +1 @@
...iga/2020/10/" data-tokens=10>10. Spieltag</option><option value=11 href="/bundesliga/2020/11/" data-tokens=11>11. Spieltag</option><option value=12 href="/bundesliga/2020/12/" data-tokens=12>12. Spieltag</option><option value=13 href="/bundesliga/2020/13/" data-tokens=13>13. Spieltag</option><option value=14 href="/bundesliga/2020/14/" data-tokens=14>14. Spieltag</option><option value=15 href="/bundesliga/2020/15/" data-tokens=15>15. Spieltag</option><option value=16 href="/bundesliga/2020/16/" data-tokens=16>16. Spieltag</option><option value=17 href="/bundesliga/2020/17/" data-tokens=17>17. Spieltag</option><option value=18 href="/bundesliga/2020/18/" data-tokens=18>18. Spieltag</option><option value=19 href="/bundesliga/2020/19/" data-tokens=19>19. Spieltag</option><option value=20 href="/bundesliga/2020/20/" data-tokens=20>20. Spieltag</option><option value=21 href="/bundesliga/2020/21/" data-tokens=21>21. Spieltag</option><option value=22 href="/bundesliga/2020/22/" data-tokens=22>22. Spieltag</option><option value=23 href="/bundesliga/2020/23/" data-tokens=23>23. Spieltag</option><option value=24 href="/bundesliga/2020/24/" data-tokens=24>24. Spieltag</option><option value=25
```
Comment (nikola-rados) [inline the_mines/data/fussballdaten_2019_2020_1.html:1]: We probably don't want to have these `html` files sitting in our repo.  Keeping 1 or 2 abbreviated files in `tests/` may be a good idea for testing some of the features here.

File: the_mines/get_html.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    return filename
+
+
+def main():
+    get_html()
+
+
+if __name__ == "__main__":
+    main()
```
Comment (nikola-rados) [inline the_mines/get_html.py]: If we are just calling `get_html()` right away I think we could just skip the call to `main()`.

```python3
if __name__ == "__main__":
    get_html()
```

File: the_mines/get_html.py [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+def get_html():
+    parser = ArgumentParser()
+    parser.add_argument(
+        "--matchday", action="store", required=True, help="ex. 8",
+    )
+    parser.add_argument(
+        "--season", action="store", required=True, help="ex. 2019/2020",
+    )
+    args = parser.parse_args()
```
Comment (nikola-rados) [inline the_mines/get_html.py]: These are all really good `args` to have.  Have you considered moving them out to before the method call (in `if __name__ == ...`)? That way we could have `matchday` and `season` as parameters.  Makes it easier to test different cases and we could call it from other modules with params.

File: the_mines/get_html.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    matchday = f"{args.matchday}"
+    season = f"{args.season}"
+
+    url = "https://www.fussballdaten.de/bundesliga/" + season[5:9] + "/" + matchday
```
Comment (nikola-rados) [inline the_mines/get_html.py:12]: It would be helpful to have a comment here explaining what the `season[5:9]` is doing.

File: the_mines/get_html.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+        print("opening file... " + filename)
+        return filename
+    except IOError:
+        # if not, then create it
+        f = open(filename, "w")
+        print("creating file..." + filename)
+        f.write(str(page_source.content))
+        shutil.move(filename, "data/" + filename)
+    finally:
+        f.close()
```
Comment (nikola-rados) [inline the_mines/get_html.py]: There are a couple things I really like here.  I appreciate your use of `try/catch/finally`, keeps things safe.  Also the use of the "early return pattern" is always nice for readability.  Lastly, thank you for closing the file!

I think it would be an even safer option to check if the file exists before trying to open it. Something like:
```python3
if not os.path.isfile(filename):
    # handle case
```
Or:
```python3
if not Path(filename).is_file():
    # handle case
```
Also, if you don't want to worry about manually closing a file, we could use a context manager:
```python3
with open("data/" + filename) as f:
    f.write(str(page_source.content))
    # and so on
```
The `with` block will automatically close the file once you leave its scope.

Would you be able to explain why we are creating and storing the file? Would it be better to just return a file object rather than just the `filename`?

File: the_mines/get_html.py (hunk 4 above)
Comment (nikola-rados) [inline the_mines/get_html.py:12]: It may also be a good idea to make the target url a parameter (we can default it to `fussballdaten`).  That makes the method more robust against future urls we may want to hit.  A change like this would probably mean having some type of url constructor method to handle known sites.

File: the_mines/get_html.py [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+    parser.add_argument(
+        "--matchday", action="store", required=True, help="ex. 8",
+    )
+    parser.add_argument(
+        "--season", action="store", required=True, help="ex. 2019/2020",
+    )
+    args = parser.parse_args()
+
+    matchday = f"{args.matchday}"
+    season = f"{args.season}"
```
Comment (nikola-rados) [inline the_mines/get_html.py]: I don't know if I understand the purpose of these lines.  Is this a way of casting the `args` to `str`?

File: the_mines/rapidapi_test.py [hunk 7]
```diff
@@ -0,0 +1,20 @@
+"""
+rapidapi requires credit card data even for free acount.
+free account gives you 50 free api calls per month and $0.25 for each call after that.
+NO THANKS!
+
+"""
```
Comment (nikola-rados) [inline the_mines/rapidapi_test.py]: This is in the wiki now, we can probably take it out of here.

File: the_mines/scrape_html.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+from argparse import ArgumentParser
+
+
+def fussballdatenpunktde_matchday_results(site_html):
+    """
+    fussballdaten.de
+    view-source:https://www.fussballdaten.de/bundesliga/2020/12/
+    """
+    # print(site_html)
+    with open(f"{site_html}") as fp:
```
Comment (nikola-rados) [inline the_mines/scrape_html.py:11]: Nice! Love context managers.

File: the_mines/scrape_html.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    print(game_dict)
+
+
+def umlaut(word_with_umlaut):
+    if "\\xc3\\xb6" in word_with_umlaut:
+        word_with_umlaut = word_with_umlaut.replace("\\xc3\\xb6", "ö")
+    elif "\\xc3\\xbc" in word_with_umlaut:
+        word_with_umlaut = word_with_umlaut.replace("\\xc3\\xbc", "ü")
+    return word_with_umlaut
+
```
Comment (nikola-rados) [inline the_mines/scrape_html.py]: Really nice catch here.  I didn't even consider there would be letters we would have to deal with.

File: the_mines/scrape_html.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    view-source:https://www.fussballdaten.de/bundesliga/2020/12/
+    """
+    # print(site_html)
+    with open(f"{site_html}") as fp:
+        soup = BeautifulSoup(fp, "html.parser")
+
+    game_count = 0
+    game_dict = {}
+    a_tag = soup("a")
+    for game_details in a_tag:
```
Comment (nikola-rados) [inline the_mines/scrape_html.py]: This is a little hard to follow, it would be helpful to comment on the different sections to explain what they are doing.

File: the_mines/scrape_html.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        required=True,
+        help="Raw HTML (ex. data/fussballdaten_matchday6.html)",
+    )
+    args = parser.parse_args()
+
+    fussballdatenpunktde_matchday_results(args)
+
+
+if __name__ == "__main__":
+    main()
```
Comment (nikola-rados) [inline the_mines/scrape_html.py]: Again, we could probably do without the `main()` method here.  Put all this stuff directly into `if __name__ ...`.

File: the_mines/run.py [hunk 12]
```diff
@@ -0,0 +1,6 @@
+from get_html import get_html
+from scrape_html import fussballdatenpunktde_matchday_results
+
+html_file = get_html()
+
+fussballdatenpunktde_matchday_results("data/" + html_file)
```
Comment (nikola-rados) [inline the_mines/run.py]: I like that you've made a run file for this.  `get_data.py` may better describe what is happening here.

File: the_mines/rapidapi_test.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+querystring = {"matchday": "1"}
+
+headers = {
+    "x-rapidapi-host": "heisenbug-bundesliga-live-scores-v1.p.rapidapi.com",
+    "x-rapidapi-key": "f62b37deecmsh70ff9936f552727p12e33fjsna71d32d2130d",
+}
+
+response = requests.request("GET", url, headers=headers, params=querystring)
+
+print(response.text)
```
Comment (nikola-rados) [inline the_mines/rapidapi_test.py]: Is this file being used anywhere?

File: the_mines/scrape_html.py [hunk 14]
```diff
@@ -0,0 +1,92 @@
+from bs4 import BeautifulSoup
+from argparse import ArgumentParser
+
+
+def fussballdatenpunktde_matchday_results(site_html):
```
Comment (nikola-rados) [inline the_mines/scrape_html.py:5]: It may be a good idea to move any `fussballdaten` specific code into a `fussballdaten/` directory.  So if we add any other websites in the future we can isolate the different scrape files.  This is something we probably don't have to worry about just yet.

File: the_mines/get_html.py [hunk 15]
```diff
@@ -0,0 +1,47 @@
...
+        f.close()
+
+    return filename
+
+
+def main():
+    get_html()
+
+
+if __name__ == "__main__":
```
Comment (nikola-rados) [inline the_mines/get_html.py]: `data_collection` may suit this directory better than `the_mines`.

**Review by nikola-rados — CHANGES_REQUESTED** [review]
Nice work, this provides a really good base for us to work off of.  I have a lot of comments, mostly styling or structural.

File: the_mines/data/fussballdaten_2019_2020_1.html (hunk 1 above)
Comment (maxUser) [reply the_mines/data/fussballdaten_2019_2020_1.html:1]: What do you mean by abbreviated?

File: the_mines/get_html.py (hunk 5 above)
Comment (maxUser) [reply the_mines/get_html.py]: Went from 11 lines to 8! Nice.

File: the_mines/get_html.py (hunk 6 above)
Comment (maxUser) [reply the_mines/get_html.py]: Not sure. Changed it to simply be `args.matchday` and `args.season`

File: the_mines/rapidapi_test.py (hunk 7 above)
Comment (maxUser) [reply the_mines/rapidapi_test.py]: I'm deleting the entire file

File: the_mines/scrape_html.py (hunk 9 above)
Comment (maxUser) [reply the_mines/scrape_html.py]: Funnily enough, this wasn't an issue when I was using `selenium` to get the `html`. The `requests` module is the issue here. After 20 minutes of failed googling for a good solution to this I just went for this dirty solution. To clarify, I prefer getting the `html` with `requests`, however.

File: the_mines/rapidapi_test.py (hunk 13 above)
Comment (maxUser) [reply the_mines/rapidapi_test.py]: No. Deleting it.

File: the_mines/get_html.py (hunk 15 above)
Comment (maxUser) [reply the_mines/get_html.py]: I respectfully disagree, _sir_

File: the_mines/get_html.py (hunk 4 above)
Comment (maxUser) [reply the_mines/get_html.py:12]: I agree. I think for now we should keep "https://www.fussballdaten.de/bundesliga/" hard-coded, but have an if statement to complete the rest of the URL depending on what stats we need.

File: the_mines/data/fussballdaten_2019_2020_1.html (hunk 1 above)
Comment (nikola-rados) [reply the_mines/data/fussballdaten_2019_2020_1.html:1]: The file is very large, we could take out the parts we aren't interested in and just keep a few cases.

File: the_mines/get_html.py (hunk 4 above)
Comment (nikola-rados) [reply the_mines/get_html.py:12]: Ok, sounds good.

File: the_mines/data/fussballdaten_2019_2020_1.html (hunk 1 above)
Comment (maxUser) [reply the_mines/data/fussballdaten_2019_2020_1.html:1]: Dunno if that's feasible. Open one and check it out - they are messy. They really aren't that large size-wise. If you can find a pain-free way to take out parts we don't need I'm all for it.

File: the_mines/scrape_html.py (hunk 14 above)
Comment (maxUser) [reply the_mines/scrape_html.py:5]: Yes, I agree

**Comment by maxUser** [comment]
@nikola-rados Is this mergeable?

File: the_mines/get_data.py [hunk 16]
```diff
@@ -0,0 +1,22 @@
...
+)
+args = parser.parse_args()
+
+
+'''
+Get data
+'''
+html_file = get_html(args.matchday, args.season)
+
+fussballdatenpunktde_matchday_results("data/" + html_file)
```
Comment (nikola-rados) [inline the_mines/get_data.py]: Adding this file to run the data collection is great! Encapsulating this section in an `if __name__ == ...` section would bring it more in line with the best practices.

File: the_mines/get_html.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+import shutil
+import os
+
+
+def get_html(matchday, season):
+
+    # - The URL of this website uses only the latter year in a season (2019/2020).
+    # - season[5:9] selects the latter season.
+    # - I'm using the 2019/2020 convention for season input because I believe it
+    #   is what most users are familiar with when asked for a season
```
Comment (nikola-rados) [inline the_mines/get_html.py:11]: Thanks for the info! You could wrap this up into a multiline comment, but I think we can save that for issue #15.

File: the_mines/get_html.py [hunk 18]
```diff
@@ -0,0 +1,34 @@
+import requests
+import shutil
+import os
+
+
+def get_html(matchday, season):
```
Comment (nikola-rados) [inline the_mines/get_html.py]: Funny things can happen when importing a method that is the same name as the module.  I would advise changing either the module name or the method name.

File: the_mines/get_html.py [hunk 19]
```diff
@@ -0,0 +1,34 @@
...
+    else:
+        print("opening file... " + filename)
+        return filename
+
+    return filename # should this be in the if statement or out here?
+
+
+if __name__ == "__main__":
+
+    get_html(matchday, season)
```
Comment (nikola-rados) [inline the_mines/get_html.py]: Since we are importing this method and using it elsewhere (rather than directly running this script) we don't need these lines.

File: the_mines/scrape_html.py [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+def umlaut(word_with_umlaut):
+    if "\\xc3\\xb6" in word_with_umlaut:
+        word_with_umlaut = word_with_umlaut.replace("\\xc3\\xb6", "ö")
+    elif "\\xc3\\xbc" in word_with_umlaut:
+        word_with_umlaut = word_with_umlaut.replace("\\xc3\\xbc", "ü")
+    return word_with_umlaut
+
+
+if __name__ == "__main__":
+    fussballdatenpunktde_matchday_results(args)
```
Comment (nikola-rados) [inline the_mines/scrape_html.py]: Same comment as above, we don't actually need this in here since we use it elsewhere.

**Review by nikola-rados — CHANGES_REQUESTED** [review]
Nice work! I appreciate all the changes you made to accommodate my recommendations. There are a few small things I think should be looked at before merging.

Two additional items that I can't comment on in the code are:
1. We probably only need a file or two of html in `data/`
2. Would you be able to add `the_mines/.DS_Store` to `.gitignore` and take it out of the repo

**Comment by maxUser** [comment]
@nikola-rados All re-request review changes have been implemented and pushed onto the branch. Let me know if I can merge.

**Comment by nikola-rados** [comment]
@maxUser merge away!

**Review by nikola-rados — APPROVED** [review]
